staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filter.go

```diff
@@ -0,0 +1,168 @@
+) Interface {
+	return &implementation{ctl: fcfc.NewTestableController(informerFactory, flowcontrolClient, serverConcurrencyLimit, requestWaitLimit, queueSetFactory)}
+}
+
+func (impl *implementation) Run(stopCh <-chan struct{}) error {
+	return impl.ctl.Run(stopCh)
+}
+
+func (impl *implementation) Wait(ctx context.Context, requestDigest fcfc.RequestDigest) (bool, func()) {
+	startWaitingTime := time.Now()
+	for { // loop until QueueSet::Wait does not return tryAnother==true
+		cfgState := impl.ctl.GetCurrentState()
+
+		// 1. classify the request
+		fsName, distinguisherMethod, plName, plEnablement, queues := cfgState.Match(requestDigest)
+
+		// 2. early out for exempt
+		if plEnablement == fctypesv1a1.PriorityLevelEnablementExempt {
+			klog.V(7).Infof("Serving requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s without delay", requestDigest.RequestInfo, requestDigest.User, fsName, plName)
+			startExecutionTime := time.Now()
+			return true, func() {
+				metrics.ObserveExecutionDuration(plName, fsName, time.Now().Sub(startExecutionTime))
+			}
+		}
+
+		// 3. computing hash
+		flowDistinguisher := computeFlowDistinguisher(requestDigest, distinguisherMethod)
+		hashValue := hashFlowID(fsName, flowDistinguisher)
+
+		// 4. queuing
+		tryAnother, execute, afterExecute := queues.Wait(ctx, hashValue, requestDigest.RequestInfo, requestDigest.User)
+		if tryAnother {
+			klog.V(5).Infof("Request requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s landed in timing splinter, re-classifying", requestDigest.RequestInfo, requestDigest.User, fsName, plName)
+			// Controller promises us that a following
+			// GetCurrentState() will produce something more evolved
+			// than last time
+			continue
+		}
+
+		// 5. execute or reject
+		metrics.ObserveWaitingDuration(plName, fsName, strconv.FormatBool(execute), time.Now().Sub(startWaitingTime))
+		if !execute {
+			klog.V(7).Infof("Rejecting requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s after fair queuing", requestDigest.RequestInfo, requestDigest.User, fsName, plName)
+			return false, func() {}
+		}
+		klog.V(7).Infof("Serving requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s after fair queuing", requestDigest.RequestInfo, requestDigest.User, fsName, plName)
+		startExecutionTime := time.Now()
+		return execute, func() {
+			metrics.ObserveExecutionDuration(plName, fsName, time.Now().Sub(startExecutionTime))
+			afterExecute()
+		}
+	}
+}
+
+// computeFlowDistinguisher extracts the flow distinguisher according to the given method
+func computeFlowDistinguisher(rd fcfc.RequestDigest, method *fctypesv1a1.FlowDistinguisherMethod) string {
+	if method == nil {
+		return ""
+	}
+	switch method.Type {
+	case fctypesv1a1.FlowDistinguisherMethodByUserType:
+		return rd.User.GetName()
+	case fctypesv1a1.FlowDistinguisherMethodByNamespaceType:
+		return rd.RequestInfo.Namespace
+	default:
+		// this line shall never reach
+		panic("invalid flow-distinguisher method")
+	}
+}
+
+var crcHashTable = crc64.MakeTable(crc64.ECMA)
```

Comment:
I still would like to switch to SHA256. The authn stack already performs at least one SHA256 operation for each request, so it is OK from a performance perspective. IMO it's really bad karma to use CRC in any context where there's even a hint of a chance of adversarial users.

Reply:
we're using 64 bits as the source of shuffle-sharding, in order to use SHA256 we will have to squash the entropy from 256 to 64 bit

Reply:
Just use the last 8 bytes? And if people ever want bigger hand/deck sizes than we currently support, it is obvious how to get more entropy.

I don't understand the pushback I'm getting on this. This is a tiny change compared to my other asks.

Reply:
I am not strongly opposed to this, and am comforted to learn that there is already one cryptographic hash being done.  I would like to do some benchmarking to quantify the effect; I expect comforting results but would like to have actual evidence.

Yes, using any 64 of the 256 bits should be good enough for us.